Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -405,23 +415,51 @@ static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&&
         RefPtr strongThis = weakThis.get();
         if (!strongThis)
             return;
-        if (error)
-            return; // FIXME: Pipe this error to JS.
+        if (error) {
+            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED))
+                strongThis->receiveDatagram({ }, false, WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            // If we were cancelled, exit gracefully
+            return;
+        }
+
+        ASSERT(content || withFin);
 
         // FIXME: Not only is this an unnecessary string copy, but it's also something that should probably be in WTF or FragmentedSharedBuffer.
         auto vectorFromData = [](dispatch_data_t content) {
-            ASSERT(content);
             Vector<uint8_t> request;
-            dispatch_data_apply_span(content, [&](std::span<const uint8_t> buffer) {
-                request.append(buffer);
-                return true;
-            });
+            if (content) {
+                dispatch_data_apply_span(content, [&](std::span<const uint8_t> buffer) {
+                    request.append(buffer);
+                    return true;
+                });
+            }
             return request;
         };
 
-        strongThis->receiveDatagram(vectorFromData(content).span());
-        strongThis->receiveDatagramLoop();
+        bool completed = !content && withFin;
+        strongThis->receiveDatagram(vectorFromData(content).span(), completed, std::nullopt);
+        if (!completed)
+            strongThis->receiveDatagramLoop();
     }).get());
 #endif // HAVE(WEB_TRANSPORT)
 }
+
+void NetworkTransportSession::terminate(WebCore::WebTransportSessionErrorCode code, CString&& message)
+{
+#if HAVE(WEB_TRANSPORT)
+    if (RetainPtr metadata = nw_connection_group_copy_protocol_metadata(m_connectionGroup.get(), adoptNS(nw_protocol_copy_webtransport_definition()).get())) {
+        nw_webtransport_metadata_set_session_error_code(metadata.get(), code);
+        nw_webtransport_metadata_set_session_error_message(metadata.get(), message.data());
+    }
+
+    if (m_datagramConnection)
+        nw_connection_cancel(m_datagramConnection.get());
+
+    for (auto& stream : m_streams.values())
+        stream->cancel(code);
+    m_streams.clear();
```

Comment:
Instead of this, you could iterate std::exchange(m_streams, { }).values() to do both at the same time.  Not too important though.